Framework/Algorithms/src/SANSSolidAngle.cpp

```diff
@@ -0,0 +1,184 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidAlgorithms/SANSSolidAngle.h"
+#include "MantidAPI/AlgorithmProperty.h"
+#include "MantidAPI/HistogramValidator.h"
+#include "MantidAPI/InstrumentValidator.h"
+#include "MantidAPI/SpectrumInfo.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidAPI/WorkspaceUnitValidator.h"
+#include "MantidDataObjects/EventList.h"
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/Logger.h"
+#include "MantidKernel/PropertyManager.h"
+#include "MantidKernel/PropertyManagerDataService.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+using namespace Kernel;
+using namespace API;
+using namespace Geometry;
+using namespace DataObjects;
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(SANSSolidAngle)
+
+void SANSSolidAngle::init() {
+
+  declareProperty(std::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input,
+                      boost::make_shared<InstrumentValidator>()),
+                  "This workspace is used to identify the instrument to use "
+                  "and also which\n"
+                  "spectra to create a solid angle for. If the Max and Min "
+                  "spectra values are\n"
+                  "not provided one solid angle will be created for each "
+                  "spectra in the input\n"
+                  "workspace");
+  declareProperty(std::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "The name of the workspace to be created as the output of "
+                  "the algorithm.  A workspace of this name will be created "
+                  "and stored in the Analysis Data Service.");
+
+  std::vector<std::string> exp_options{"Normal", "Tube", "Wing"};
+  declareProperty("Type", "Normal",
+                  boost::make_shared<StringListValidator>(exp_options),
+                  "Select the method to calculate the Solid Angle.\n"
+                  "Normal: cos^3(theta); Tube: cons(alpha)*cos^3(theta); "
```

Comment:
The formula for the tube is wrong, it also has a typo.

Reply:
@quantumsteve Can you change it to: `cos(alpha)*cos^2(theta)` ?

Reply:
The math is right, the typo is fixed.

Reply:
It must be `cosˆ2`